public/app/plugins/datasource/influxdb/components/VisualInfluxQLEditor/Editor.tsx

```diff
@@ -0,0 +1,221 @@
+      // probably because this way multi-value variables work ok too.
+      .map((v) => `/^$${v.name}$/`)
+  );
+}
+
+// helper function to make it easy to call this from the widget-render-code
+function withTemplateVariableOptions(optionsPromise: Promise<string[]>): Promise<string[]> {
+  return optionsPromise.then((options) => [...getTemplateVariableOptions(), ...options]);
+}
+
+const SectionWrap = ({ initialName, children }: { initialName: string; children: React.ReactNode }) => (
+  <div className="gf-form-inline">
+    <SectionLabel name={initialName} isInitial={true} />
+    {children}
+    <SectionFill />
+  </div>
+);
+
+export const Editor = (props: Props): JSX.Element => {
+  const query = normalizeQuery(props.query);
+  const { datasource } = props;
+  const onAppliedChange = (newQuery: InfluxQuery) => {
+    props.onChange(newQuery);
+    props.onRunQuery();
+  };
+  const handleFromSectionChange = (policy: string | undefined, measurement: string | undefined) => {
+    onAppliedChange({
+      ...query,
+      policy,
+      measurement,
+    });
+  };
+
+  const handleTagsSectionChange = (tags: InfluxQueryTag[]) => {
+    // we set empty-arrays to undefined
+    onAppliedChange({
+      ...query,
+      tags: tags.length === 0 ? undefined : tags,
+    });
+  };
+
+  const dynamicSelectPartOptions = new Map([
+    ['field_0', () => getFieldKeysForMeasurement(unwrap(query.measurement), query.policy, datasource)],
+  ]);
+  const selectLists = (query.select ?? []).map((sel) => makePartList(sel, dynamicSelectPartOptions));
+
+  const dynamicGroupByPartOptions = new Map([
+    [
+      'tag_0',
+      () => getTagKeysForMeasurementAndTags(unwrap(query.measurement), query.policy, query.tags ?? [], datasource),
+    ],
+  ]);
+
+  const groupByList = makePartList(query.groupBy ?? [], dynamicGroupByPartOptions);
+
+  return (
+    <div>
+      <SectionWrap initialName="from">
+        <FromSection
+          policy={query.policy}
+          measurement={query.measurement}
+          getPolicyOptions={() => getAllPolicies(datasource)}
+          getMeasurementOptions={() => withTemplateVariableOptions(getAllMeasurements(datasource))}
+          onChange={handleFromSectionChange}
+        />
+        <SectionLabel name="where" />
+        <TagsSection
+          tags={query.tags ?? []}
+          onChange={handleTagsSectionChange}
+          getTagKeyOptions={() =>
+            getTagKeysForMeasurementAndTags(unwrap(query.measurement), query.policy, query.tags ?? [], datasource)
+          }
+          getTagValueOptions={(key: string) =>
+            withTemplateVariableOptions(getTagValues(key, unwrap(query.measurement), query.policy, datasource))
+          }
+        />
+      </SectionWrap>
+      {selectLists.map((sel, index) => (
+        <SectionWrap key={index.toString()} initialName={index === 0 ? 'select' : ''}>
+          <PartListSection
+            key={index.toString()}
```

Comment:
This key is probably not needed (?) - only the first element needs it (`SectionWrap` above). Also I think it doesn't need to be a string, `key={index}` should be enough.

Reply:
right on both :)